packages/backend/src/models/ProjectModel/ProjectModel.ts

```diff
@@ -569,57 +570,121 @@ export class ProjectModel {
-            explore = await this.getExploreQueryBuilder(
-                projectUuid,
-            ).andWhereRaw("(explore->>'tables')::json->? IS NOT NULL", [
-                tableName,
-            ]);
-        }
+        return wrapOtelSpan(
+            'ProjectModel.findExploreByTableName',
+            {},
+            async (span) => {
+                // check individually cached explore
+                let exploreCache = await this.database(CachedExploreTableName)
+                    .select('explore')
+                    .whereRaw('? = ANY(table_names)', tableName)
+                    .andWhere('project_uuid', projectUuid)
+                    .first();
+
+                span.setAttribute(
+                    'foundIndividualExploreCache',
+                    !!exploreCache,
+                );
+                if (!exploreCache) {
+                    // fallback: check all cached explores
+                    // try finding explore via base table name
+                    exploreCache = await this.getExploreQueryBuilder(
+                        projectUuid,
+                    ).andWhereRaw("explore->>'baseTable' = ?", [tableName]);
+
+                    if (!exploreCache) {
+                        // try finding explore via joined table alias
+                        // Note: there is an edge case where we return the wrong explore because join table aliases are not unique
+                        exploreCache = await this.getExploreQueryBuilder(
+                            projectUuid,
+                        ).andWhereRaw(
+                            "(explore->>'tables')::json->? IS NOT NULL",
+                            [tableName],
+                        );
+                    }
+                }
 
-        return explore
-            ? ProjectModel.convertMetricFiltersFieldIdsToFieldRef(
-                  explore.explore,
-              )
-            : undefined;
+                return exploreCache
+                    ? ProjectModel.convertMetricFiltersFieldIdsToFieldRef(
+                          exploreCache.explore,
+                      )
+                    : undefined;
+            },
+        );
     }
 
     async saveExploresToCache(
         projectUuid: string,
         explores: (Explore | ExploreError)[],
     ): Promise<DbCachedExplores> {
-        const [cachedExplores] = await this.database(CachedExploresTableName)
-            .insert({
-                project_uuid: projectUuid,
-                explores: JSON.stringify(explores),
-            })
-            .onConflict('project_uuid')
-            .merge()
-            .returning('*');
-        return cachedExplores;
+        return wrapOtelSpan(
+            'ProjectModel.saveExploresToCache',
+            {},
+            async () => {
+                // cache explores individually
+                await this.database(CachedExploreTableName)
```

Comment:
Doesn't this change the existing behaviour.

If I previously pushed 3 explores. Then afterwards I push just 2 explores. Don't I still end up with all 3 explores? 

You should delete all the explores before inserting the new ones. So if the user pushes a new project with fewer explores, they'll be removed from the database.

Maybe I've missed something.